drivers/platform/pico/pico_irq.c

```diff
@@ -0,0 +1,442 @@
+ * @return -ENOSYS
+ */
+int32_t pico_trigger_level_set(struct no_os_irq_ctrl_desc *desc,
+			       uint32_t irq_id,
+			       enum no_os_irq_trig_level level)
+{
+	return -ENOSYS;
+}
+
+/**
+ * @brief Register a callback.
+ * @param desc   - The IRQ controller descriptor.
+ * @param irq_id - Interrupt identifier.
+ * @param cb     - Descriptor of the callback.
+ * @return 0 if successful, error code otherwise.
+ */
+int32_t pico_irq_register_callback(struct no_os_irq_ctrl_desc *desc,
+				   uint32_t irq_id,
+				   struct no_os_callback_desc *cb)
+{
+	int ret;
+	struct irq_action *action;
+	struct irq_action action_key = {.irq_id = irq_id};
+
+	switch (cb->peripheral) {
+	case NO_OS_UART_IRQ:
+		/* Set up the interrupt handler */
+		irq_set_exclusive_handler(irq_id,
+					  cb->handle == uart0 ? on_uart0_rx : on_uart1_rx);
+
+		if (_events[cb->event].actions == NULL) {
+			ret = no_os_list_init(&_events[cb->event].actions,
+					      NO_OS_LIST_PRIORITY_LIST,
+					      irq_action_cmp);
+			if (ret) {
+				irq_remove_handler(irq_id,
+						   cb->handle == uart0 ? on_uart0_rx : on_uart1_rx);
+				return ret;
+			}
+		}
+		break;
+	case NO_OS_TIM_IRQ:
+		/* Set up the interrupt handler */
+		hardware_alarm_set_callback(irq_id, _alarm_callback);
+		/* By default, disable alarm irq.
+		The user shall enable the interrupt when seen fit. */
+		irq_set_enabled(irq_id, false);
+
+		if (_events[cb->event].actions == NULL) {
+			ret = no_os_list_init(&_events[cb->event].actions,
+					      NO_OS_LIST_PRIORITY_LIST,
+					      irq_action_cmp);
+			if (ret) {
+				hardware_alarm_set_callback(irq_id, NULL);
+				return ret;
+			}
+		}
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	ret = no_os_list_read_find(_events[cb->event].actions,
+				   (void **)&action,
+				   &action_key);
+	/*
+	 * If an action with the same irq_id as the function parameter does not exists, insert a new one,
+	 * otherwise update
+	 */
+	if (ret) {
+		action = calloc(1, sizeof(*action));
+		if (!action)
+			return -ENOMEM;
+
+		action->irq_id = irq_id;
+		action->handle = cb->handle;
+		action->callback = cb->callback;
+		action->ctx = cb->ctx;
+
+		ret = no_os_list_add_last(_events[cb->event].actions, action);
+		if (ret) {
+			free(action);
+			return ret;
+		}
+	} else {
+		action->irq_id = irq_id;
+		action->handle = cb->handle;
+		action->callback = cb->callback;
+		action->ctx = cb->ctx;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Unregister a callback.
+ * @param desc   - Interrupt controller descriptor.
+ * @param irq_id - Id of the interrupt.
+ * @param cb     - Descriptor of the callback.
+ * @return 0 if successful, error code otherwise.
+ */
+int32_t pico_irq_unregister_callback(struct no_os_irq_ctrl_desc *desc,
+				     uint32_t irq_id, struct no_os_callback_desc *cb)
+{
+	int ret;
+	void *discard;
+	struct irq_action key;
+
+	switch (cb->peripheral) {
+	case NO_OS_UART_IRQ:
+		break;
+	case NO_OS_TIM_IRQ:
+		break;
+	default:
+		return -EINVAL;
+	}
```

Comment:
Is this necessary? Your `register_callback` function will fail if other peripherals (than these 2) are used.

Reply:
not necessary, thanks!